### PR TITLE
Proposed clarifications

### PR DESCRIPTION
**Pull request by halostatue**

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # Bru Lang
 
-Bru is a simple markup language with json like semantics.
+Bru is a simple markup language with [JSON](https://json.org)-like semantics.
```
Comment (halostatue) [inline readme.md:3]: This points people to the main syntactical definition of JSON, for which Bru will be leaning on the *meaning* of most supported primitive types. That is, a Bru `number` is the same as a JSON `number`.

File: readme.md [hunk 2]
```diff
@@ -30,6 +30,11 @@ http: {
 * Indentation based syntax
```
Comment (halostatue) [inline readme.md:30]: I didn't change this, but typically whitespace indentation languages do not use delimiters. I would remove this as a design goal even if it is *desirable* that within composite types the primitive values are indented for readability.

File: readme.md [hunk 3]
```diff
@@ -30,6 +30,11 @@ http: {
...
 * Annotations for providing additional information
 
+The top level of a `.bru` file is an implicit Multimap. It does not require
```
Comment (halostatue) [inline readme.md:33]: This was added because one would expect the `http:` in the example to be inside a `{}` block. It is not, which means that there is an implicit object in place.

File: readme.md [hunk 4]
```diff
@@ -30,6 +30,11 @@ http: {
...
+or support braces.
+
+Except where noted, blank lines and whitespace are ignored.
```
Comment (halostatue) [inline readme.md:36]: This is related to the comment above about whitespace indentation, but indicates that although there is whitespace, it is *not* included except where noted.

File: readme.md [hunk 5]
```diff
@@ -39,7 +44,10 @@ There are 4 primitive types in Bru.
 * Boolean
 * Null
 
-A string can contain any character except `:` and `\n`
+The string type slightly differs from JSON in that a Bru string is
+a single-quoted and is always UTF-8. It may contain any printable UTF-8
+character except `'` or `\n`. Whitespace is considered significant inside of the
+quoted string.
```
Comment (halostatue) [inline readme.md:50]: I changed this for multiple reasons:

- examples string values include `:` as part of URLs, so it cannot be true that they cannot contain `:`
- restricting to printable characters makes the behaviour *generally* correct for HTTP headers, query parameters, etc.

I am less comfortable with the quoting rules, but I do not know enough about the Bru language parser to know whether you support escapes (e.g., `'it\'s awesome'`) in strings or whether single quotes are the only supported quotes (`"it's awesome"`). If possible, I would consider copying the [hjson](https://hjson.github.io) rules on strings (some strings can be unquoted).

File: readme.md [hunk 6]
```diff
@@ -48,11 +56,9 @@ There are 3 composite types in Bru.
 * [Multiline String](https://en.wikipedia.org/wiki/Here_document)
 
 #### Multimap
-Multimap is essentially a dictionary (key-value pair) that can have duplicate keys.
-Its enclosed in curly braces `{}`.
-Keys and Values are separated by a colon `:` and key-value pairs are separated by a newline `\n`.
+Multimap is a dictionary (key-value pair) with duplicate keys, enclosed in curly braces (`{}`). Keys and Values are separated by a colon `:` and key-value pairs are separated by a newline (`\n`).
 
-All keys are strings. Values can be primitive types or composite types.
+All keys are unquoted strings and must not contain spaces. Values may be primitive or composite types. Some contexts may offer further restrictions not required by the Bru language.
```
Comment (halostatue) [inline readme.md:61]: It's likely that we want to be more specific about this. As I noted, "some contexts" are more restrictive on key names (e.g., headers *must* be legal HTTP header name values, so `[-a-zA-Z0-9]+`).

File: readme.md [hunk 7]
```diff
@@ -65,7 +71,7 @@ All keys are strings. Values can be primitive types or composite types.
 ```
 
 #### Array
-Array is a list of values separated by a newline `\n`. Its enclosed in square brackets `[]`.
+Array is a list of values separated by a newline (`\n`), enclosed in square brackets (`[]`).
```
Comment (halostatue) [inline readme.md:74]: Should this be `primitive` values, or would it be possible to have a list of multimaps?

File: readme.md [hunk 8]
```diff
@@ -74,7 +80,12 @@ tags: [
 ```
 
 #### Multiline String
-Multiline string is a string that spans multiple lines. Its enclosed in parenthesis `()`.
+Multiline string is a string that spans multiple lines, enclosed in parentheses (`()`). The
+content *must* begin on the line after the opening parenthesis (`(`) and the
+closing parenthesis (`)`) *must* be on the line after the end of the multiline
+string. Multiline strings are indented at least two spaces more than the key
+which contains them. This indentation is removed on parsing.
```
Comment (halostatue) [inline readme.md:87]: If it is possible to do, using a more typical multiline character approach (`'''` or `"""` delimited like python, elixir, or hjson) would be better than parentheses. However, I added much more explicit definition on where the text begins (the line *after* opening parenthesis) and ends (the line *before* closing parenthesis) and what happens to the indented text.

Is the indented text *always stripped at `key-indent + 2` or is it `first non-whitespace of first line`?

That is, is:

```
(
  xyz
)
```

the same as:

```
(
      xyz
)
```

It may be desirable to add an *explicit* optional boundary character if there should be a difference:

```
(
      |   xyz
)
```

In this case, the value would be `"   xyz"`.

Also unclear: does the *last* line of a multiline string include a newline? That is, if I want to have `abc\ndef\n`, do I do:

```
(
  abc
  def
)
```

or

```
(
  abc
  def

)
```

File: readme.md [hunk 9]
```diff
@@ -85,10 +96,15 @@ article: {
...
+```text
+Bru is a simple markup language with json like semantics.
+It's currently used in Bruno to save details of an api request in a file.
+```
+
 ### Annotations
-Annotations are used to provide additional information about a key-value pair.
-An annotation starts with `@` and ends with a newline `\n`.
-Optionally, Comma seperated args can be passed to an annotation.
+Annotations are used to provide additional information about a key-value pair. An annotation starts with (`@`) and ends with a newline (`\n`). Arguments may be passed to an annotation in parentheses (`()`) and separated by commas. Argument values may *only* be primitive types.
```
Comment (halostatue) [inline readme.md:107]: Making assumptions here, but there is no indication that annotation with arguments could span multiple lines. Should they be able to do so?

**Review by halostatue — COMMENTED** [review]
I’ve added some comments to express my thinking on this.

File: readme.md (hunk 1 above)
Comment (helloanoop) [reply readme.md:3]: 👍

File: readme.md (hunk 2 above)
Comment (helloanoop) [reply readme.md:30]: Hmm.. Let's discuss more this on a call (will schedule one soon)

In general, given that we want to support annotations, indentation along delimiters (`{}`) makes the format more readable

Though the below example will give an impression that the one without delimiters look clean and simple, I think when you want to support arrays and multiline strings - a consistent delimiter structure approach of multimap (`{}`),  arrays (`[]`) and multiline strings (`()`) would be better

Would love to discuss further in a call.

#### Without delimiters
```groovy
http:
  method: 'GET'
  url: 'https://www.usebruno.com/hello'
  headers:
    Content-Type: 'application/json'

    @disabled
    @description('This is a sample request')
    Authorization: 'Bearer{{token}}'
  param:
    query:
      @description('The status of the user')
      @enum('active', 'inactive')
      status: 'active'
```

#### With delimiters 
```groovy
http: {
  method: 'GET'
  url: 'https://www.usebruno.com/hello'
  headers: {
    Content-Type: 'application/json'

    @disabled
    @description('This is a sample request')
    Authorization: 'Bearer{{token}}'
  }
  param: {
    query: {
      @description('The status of the user')
      @enum('active', 'inactive')
      status: 'active'
    }
  }
```

PS: The above example does not take into consideration of how an indentation based approach without delimiters would work in case of arrays and multi-line strings

File: readme.md (hunk 3 above)
Comment (helloanoop) [reply readme.md:33]: 👍

File: readme.md (hunk 4 above)
Comment (helloanoop) [reply readme.md:36]: 👍

File: readme.md (hunk 5 above)
Comment (helloanoop) [reply readme.md:50]: > restricting to printable characters makes the behaviour generally correct for HTTP headers, query parameters, etc.

We would not be able to restrict the contents to just UTF-8 as we also have to support other languages. Ex: see https://github.com/usebruno/bruno/issues/671 in key value pairs

> but I do not know enough about the Bru language parser to know whether you support escapes

Yes, we should support escapes

File: readme.md (hunk 6 above)
Comment (helloanoop) [reply readme.md:61]: > All keys are unquoted strings and must not contain spaces.

I think we should allow all characters in a key. Here is how I am thinking
* They key have any unicode character expect (`:`), if it has `:`, then it must be eccaped
* The key cannot be surrounded by whitespaces
* The surrounding whitespaces are trimmed while parsing.

@halostatue I think a live playground with a parser transforming `bru` to `json` would convey my motivation here. Will let you know when I have that up and running. 

Again, This is not set in stone. Would love to discuss nuances in a call.

File: readme.md (hunk 7 above)
Comment (helloanoop) [reply readme.md:74]: Yes, This should possible. Here is an example - 

```groovy
person: {
  name: 'Bruno'
  links: [
    {
      site: 'github'
      url: 'https://www.github.com/usebruno/bruno'
    }
    {
      site: 'twitter'
      url: 'https://www.twitter.com/use_bruno'
      description: (
        We are on twitter.
        Follow us to get updates about Bruno.
      )
    }
  ]
}
```

File: readme.md (hunk 8 above)
Comment (helloanoop) [reply readme.md:87]: So based on my reading the delimter format was `"""` was introduced primarily in bash shells, and has been adopted across programming languages.

There has been confusion around how new line gets treated.
For instance, TOML assumes the sentence begins in the next line after `'''`, but does not trim the last line

#### TOML
`text`
```text
abc
def
```

`toml`
```toml
data = '''
abc
def'''
```

not so clean if you don't want your data to have a new line

#### YAML
YAML is even more confusing and there appears to be [9 ways](https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-in-yaml-over-multiple-lines) of how to represent multiline strings 

#### Bru
With Bru we have an opportunity to make this very simple.
You put the rules very succinctly
* content *must* begin on the line after the opening parenthesis (`(`) and the
closing parenthesis (`)`) *must* be on the line after the end of the multiline
string.
* Multiline strings are indented *two spaces* more than the key
which contains them. This indentation is removed on parsing.

#### Some examples in Bru
`text`
```text
line 1
line 2
```

`bru`
```groovy
data: (
  line 1
  ine 2
)
```

The below example has a newline before and after
`text`
```text

  line 1 with 2 spaces before
line 2
     line 3 with 4 spaces before

```

`bru`
```groovy
data: (
  
    line 1 with 2 spaces before
  line 2
      line 3 with 4 spaces before
  
)
```

On the choice of `()` as delimiters, I feel this makes the format a tad bit less noisy and more elegant.
IMO, we don't need three `'''` quotes to delimit a multiline string when we the language is indentation based and specifies clearly how the multiline string gets parsed.

It also plays well in terms of consistency of having brackets as delimters for all the composite types
* multimap `{}`
* array - `[]`
* multistring - `()`

We can discuss further on this in the call.

File: readme.md (hunk 9 above)
Comment (helloanoop) [reply readme.md:107]: IMO, Although we could potentially have the lang allow composite types as args to annotations, we should keep it simple right now and just have **only** support primitive types as args

In the future, if a strong need arises, then it should be possible to extend it like below.
It would be by allowing args to be delimited by newlines. 
```
{
  query: {
    @description('The status of the user')
    @enum(
      {
        name: 'active'
        description: 'The user is active'
      }
      {
        name: 'inactive'
        description: 'The user is inactive'
      }
    )
    status: active
  }
}
```

File: readme.md (hunk 2 above)
Comment (halostatue) [reply readme.md:30]: I prefer *with* delimiters (it also makes parsing much easier); I am recommending that indentation not be identified as syntactically significant (except where noted for multi-line strings) but preferred for readability.

File: readme.md (hunk 5 above)
Comment (halostatue) [reply readme.md:50]: Does JavaScript have good support or non-UTF-8 encoded languages? I know that for a while, the JS string was (is?) UCS-2, which is its own nightmare.

If non-UTF-8 encoding can be supported, I would recommend that some sort of annotation be used. Maybe something like `r'raw string that might be non-utf-8'` as Python does. If not an annotation, perhaps a different delimiter set, like `<<raw string that might not be non-utf-8>>`. Multi-line strings would need a similar distinction `r(\n…\n)` or `<<(\n…\n)>>`.

File: readme.md (hunk 6 above)
Comment (halostatue) [reply readme.md:61]: Or keys with special characters (whitespace, `:`, etc.) must be quoted. I was trying to infer some rules based on what I had seen. I would keep a variation of the `some contexts…` comment, as header keys *must* conform to a very specific ruleset.

File: readme.md (hunk 7 above)
Comment (halostatue) [reply readme.md:74]: Which also means that nested arrays are allowed. Worth calling out in the syntax description.

File: readme.md (hunk 8 above)
Comment (halostatue) [reply readme.md:87]: As far as I know, bash does not support `'''` or `"""` as special forms. Python would be the language that I am aware of that introduced this, and they have always been multi-line.

Wording the indentation spec here is going to be subtle, because not all multi-lines will be attached to keys. I would reword that to something like:

> Multi-line strings are enclosed with `(` and `)`, which must be on separate lines from the content of the multi-line string. The content must be indented two spaces *more* than the beginning of the line where the opening parenthesis begins.
>
> Examples:
>
> ```groovy
> key: (
>   content starts here
> )
> ```
>
> ```groovy
> [
>   (
>     content starts here
>   )
> ]

I realize that interoperability with other tools is not a primary goal, but if the spec is clear enough to allow the implementation of non-Bruno parsers, especially in other languages, then there will be some comfort that the spec is robust.

**Comment by helloanoop** [comment]
@halostatue I am going to be merge this as I am mostly in agreement with all the things you mentioned. Also, your PR adds specificity to the spec and should be helpful to anyone who is visiting the repo at this stage

These are the things I feel we need to re-review 
1. UTF8 only strings in key and value
2. Delimiter for multi line strings
3. Whether indentation strictness should be limited just to multiline strings.

Its fine as this is still in a draft stage.

Next rough steps I am thinking
- `bru-js` parser library
- Live playground
- Community Call to review
- If all goes well, Ratify a Release Candidate
- Integrate the lang with Bruno with a set of closed group and test
- If all goes well, Tag v1.0
